File: 02 백준/03 그래프 탐색/Java/src/Prob7569.java

/**
* title  : 7569. 토마토 (Gold 5)
* time   : 656ms
* memory : 126032KB
*/

import java.io.*;
import java.util.*;

public class Prob7569 {
    
    static int H, R, C;
    static int[][][] board;

    static int[] dh = {1, -1, 0, 0, 0, 0};
    static int[] dr = {0, 0, 0, 0, -1, 1};
    static int[] dc = {0, 0, 1, -1, 0, 0};

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        C = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());

        board = new int[H][R][C];
        Queue<int[]> q = new LinkedList<>();

        int tomato = 0;
        for (int h = 0; h < H; h++) {
            for (int r = 0; r < R; r++) {
                st = new StringTokenizer(br.readLine());
                for (int c = 0; c < C; c++) {
                    board[h][r][c] = Integer.parseInt(st.nextToken());

                    if (board[h][r][c] == 1) {
                        q.add(new int[] {h, r, c});
                    } else if (board[h][r][c] == 0) {
                        tomato++;
                    }
                }
            }
        }

        System.out.println(bfs(q, tomato));
    }

    public static int bfs(Queue<int[]> q, int tomato) {

        int answer = -1;
        while (!q.isEmpty()) {

            answer++;

            int length = q.size();
            for (int i = 0; i < length; i++) {

                int[] now = q.poll();

                for (int d = 0; d < 6; d++) {
                    int nh = now[0] + dh[d];
                    int nr = now[1] + dr[d];
                    int nc = now[2] + dc[d];

                    if (nh < 0 || nh >= H || nr < 0 || nr >= R || nc < 0 || nc >= C) continue;
                    if (board[nh][nr][nc] == 0) {
                        q.add(new int[] {nh, nr, nc});
                        board[nh][nr][nc] = 1;
                        tomato--;
                    }
                }
            }
        }
        return (tomato > 0) ? -1 : answer;
    }
}

File: 02 백준/03 그래프 탐색/Python/Prob7569.py

"""
title  : 7569. 토마토 (Gold 5)
time   : 1788ms     852ms
memory : 48420KB    217720KB
"""

import sys
from collections import deque

def solution():

    input = sys.stdin.readline
    C, R, H = map(int, input().split())     # 상자의 크기 R, C, 상자의 수 H

    not_tomato = 0
    tomato = deque()

    box = [[] for _ in range(H)]
    for h in range(H):
        for r in range(R):
            box[h].append(list(map(int, input().split())))     # 1 익은 토마토, 0 익지 않은 토마토, -1 비어있음
            for c in range(C):
                if box[h][r][c] == 0:
                    not_tomato += 1
                elif box[h][r][c] == 1:
                    tomato.append((h, r, c))

    return move_tomato(R, C, H, box, tomato, not_tomato)
    
def move_tomato(R, C, H, box:list, tomato:deque, not_tomato:int):

    dir = [[1, 0, 0], [-1, 0, 0], [0, -1, 0], [0, 1, 0], [0, 0, -1], [0, 0, 1]]

    day = -1

    while tomato:

        day += 1

        for _ in range(len(tomato)):

            h, r, c = tomato.popleft()

            for d in range(6):

                nh = h + dir[d][0]
                nr = r + dir[d][1]
                nc = c + dir[d][2]

                if nh < 0 or nh >= H or nr < 0 or nr >= R or nc < 0 or nc >= C or box[nh][nr][nc] == -1 or box[nh][nr][nc] == 1:
                    continue

                box[nh][nr][nc] = 1
                tomato.append((nh, nr, nc))

                not_tomato -= 1

    return -1 if not_tomato else day

print(solution())